### PR TITLE
Refactor `FrameRefType` and add `Clear` reference type

### PR DESCRIPTION
**Pull request by bjoeris**
This change simplifies the `FrameRefType` enum by removing the separate states and
transitions, and replacing them with a set of reference types, together
with functions that compose those reference types.

This change also introduces a `Clear` reference type, which is used to represent
an access to the resource that completely overwrites the previous contents; this
is in contrast with a `Write` access that may leave some of the previous
contents untouched.

This change has been tested on Vulkan, but not with the other APIs. I am a little nervous that this may break something in another API.

### CODE REVIEW

File: renderdoc/core/resource_manager.h
Comment (baldurk) [inline renderdoc/core/resource_manager.h:103]: It would be helpful to ASCII-art or comment in prose here the state machine you're constructing, as reference.

File: renderdoc/core/resource_manager.cpp
Comment (baldurk) [inline renderdoc/core/resource_manager.cpp]: I think `if((first == eFrameRef_Read) != (second == eFrameRef_Read))` should generate identical assembly and IMO is a bit clearer about the intention here.

File: renderdoc/core/resource_manager.cpp
Comment (baldurk) [inline renderdoc/core/resource_manager.cpp]: I think I've convinced myself that this case is fine but it could maybe do with a bit of rephrasing on the comment. If I follow your intentions correctly then "[it] will finish in Clear state" is a bit misleading because we're composing these to find the _starting_ state. i.e. a `ReadBeforeWrite` followed by a `Clear` sounds like it would 'finish' in a `Clear`, but we're thinking in the opposite direction to find out its first use in the frame, so we'd consider the composition to be 'it was first used as `ReadBeforeWrite`'. Instead we're trying to indicate that a partial write followed by a complete write, without any intervening or proceeding reads, is as good as a complete write.

This might also be covered by a comment in the header explaining exactly what these states are going to mean going forward.

File: renderdoc/core/resource_manager.cpp
Comment (baldurk) [inline renderdoc/core/resource_manager.cpp:144]: Prefer `{}` braces around all cases in an if/else if they're present around any one.

**Review by baldurk — CHANGES_REQUESTED** [review]
This looks good, I've made some comments mostly only wanting further clarification on exactly what each state means and how they're composed. I fully admit my fault that this wasn't documented well before but since we're changing it and since it will be a bit more involved/impactful now it'd be good to clarify what everything means. You've also got a couple of CI issues to sort out.

In terms of testing other APIs, I can take care of that. I don't think this should be a problem though - the only bit of code that used to be using this was the D3D11 code that would skip initial contents if they were large and it wasn't in `eFrameRef_ReadBeforeWrite` state. Since I removed that code in v1.2 I don't think anything cares as long as it can tell read-only access from write access, which you're preserving just fine.

**Comment by bjoeris** [comment]
I think I have addressed all comments, and attempted to fix the CI issue.

File: renderdoc/core/resource_manager.h
Comment (baldurk) [inline renderdoc/core/resource_manager.h]: typo 'colud'

File: renderdoc/core/resource_manager.h
Comment (baldurk) [inline renderdoc/core/resource_manager.h]: typo 'initila'

File: renderdoc/core/resource_manager.h
Comment (baldurk) [inline renderdoc/core/resource_manager.h]: typo 'resuorce'

File: renderdoc/core/resource_manager.h
Comment (baldurk) [inline renderdoc/core/resource_manager.h]: typo 'unkown'

**Review by baldurk — APPROVED** [review]
I highlighted a couple of typos but otherwise these comments are perfect, thank you very much. Approved now pending a final CI fix.

**Comment by bjoeris** [comment]
I have fixed the typos, attempted to address the CI error, and squashed down the intermediate commit. Assuming CI passes this time I think it's good to go.

**Comment by baldurk** [comment]
Looks like there's still a clang-format issue identified on travis.

**Pull request merged**
